test/distributed/org/apache/cassandra/distributed/test/sai/ConcurrencyFactorTest.java

```diff
@@ -0,0 +1,145 @@
+
+        insertRows(1_000_000_000L, 16_000_000_000L, 1_000_000_000L);
+
+        // flush all nodes, expected row distribution by partition key value
+        // node0: 9, 14, 12, 3
+        // node1: 5, 10, 13, 11, 1, 8, 0, 2
+        // node2: 4, 15, 7, 6
+        cluster.forEach((node) -> node.flush(KEYSPACE));
+
+        // We are expecting any of 3 specific trace messages indicating how the query has been handled:
+        //
+        // Submitting range requests on <n> ranges with a concurrency of <n>
+        //   Initial concurrency wasn't estimated and max concurrency was used instead (and SAI index was involved)
+        // Submitting range requests on <n> ranges with a concurrency of <n> (<m> rows per range expected)
+        //   Initial concurrency was estimated based on estimated rows per range (non-SAI range query)
+        // Executing single-partition query on <table>
+        //   Non-range single-partition query
+
+        // SAI range query so should bypass initial concurrency estimation
+        String query = String.format("SELECT state FROM %s.%s WHERE gdp > ? AND gdp < ? LIMIT 20", KEYSPACE, SAI_TABLE);
+        runAndValidate("Submitting range requests on 3 ranges with a concurrency of 3", query, 3_000_000_000L, 7_000_000_000L);
+
+        // Partition-restricted query so not a range query
+        query = String.format("SELECT state FROM %s.%s WHERE pk = ?", KEYSPACE, SAI_TABLE);
+        runAndValidate("Executing single-partition query on sai_simple_primary_key", query, 0);
+
+        // Token-restricted range query not using SAI so should use initial concurrency estimation
+        query = String.format("SELECT * FROM %s.%s WHERE token(pk) > 0", KEYSPACE, SAI_TABLE);
+        runAndValidate("Submitting range requests on 2 ranges with a concurrency of 2 (230.4 rows per range expected)", query);
+
+        // Token-restricted range query with SAI so should bypass initial concurrency estimation
+        query = String.format("SELECT * FROM %s.%s WHERE token(pk) > 0 AND gdp > ?", KEYSPACE, SAI_TABLE);
+        runAndValidate("Submitting range requests on 2 ranges with a concurrency of 2", query, 3_000_000_000L);
+    }
+
+    /**
+     * Run the given query and check that the given trace message exists in the trace entries.
+     */
+    private void runAndValidate(String trace, String query, Object... bondValues)
+    {
+        UUID sessionId = TimeUUID.Generator.nextTimeAsUUID();
+
+        cluster.coordinator(1).executeWithTracingWithResult(sessionId, query, ConsistencyLevel.ALL, bondValues);
+
+        await().atMost(5, TimeUnit.SECONDS).until(() -> {
+            List<TracingUtil.TraceEntry> traceEntries = TracingUtil.getTrace(cluster, sessionId, ConsistencyLevel.ONE);
+            return traceEntries.stream().anyMatch(entry -> entry.activity.equals(trace));
+        });
+    }
+
+    private TokenSupplier generateTokenSupplier()
+    {
+        List<String>[] tokens = new List[3];
```

Comment:
Nit: We could get rid of the unchecked cast warning with either `@SuppressWarning` or maybe by using a list:
```java
List<List<String>> tokens = Arrays.asList(Collections.singletonList("-9223372036854775808"),
                                          Collections.singletonList("-3074457345618258602"),
                                          Collections.singletonList("3074457345618258603"));
return nodeIdx -> tokens.get(nodeIdx - 1);
```